macros/src/ast_node.rs

```diff
@@ -0,0 +1,172 @@
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote, ToTokens};
+use std::collections::HashMap;
+use syn::{parse_quote, punctuated::Punctuated, Fields, Ident, ItemEnum, Token};
+
+pub struct State<'a> {
+    enums: &'a [ItemEnum],
+    tree: HashMap<&'a Ident, Vec<&'a Ident>>,
+}
+
+impl<'a> State<'a> {
+    pub fn new(enums: &'a [ItemEnum]) -> Self {
+        Self {
+            enums,
+            tree: enums
+                .iter()
+                .map(|e| (&e.ident, e.variants.iter().map(|v| &v.ident).collect()))
+                .collect(),
+        }
+    }
+
+    pub fn generate_enum_definitions(&self, stream: &mut TokenStream) {
+        for flavor in FLAVORS {
+            for definition in self.enums {
+                let new_def = flavor.transform(definition, |x| self.tree.contains_key(x));
+                new_def.to_tokens(stream);
+            }
+        }
+    }
+
+    pub fn generate_transitive_from_impls(&self, stream: &mut TokenStream) {
+        let root = &self.enums[0].ident;
+
+        for category in &self.tree[root] {
+            for node in &self.tree[category] {
+                if let Some(leaves) = self.tree.get(node) {
+                    // From<KeywordEnum> for NodeEnum
+                    self.emit_from_impl([node, category, root], true, stream);
+
+                    for leaf in leaves {
+                        // From<KeywordTime> for LeafEnum
+                        self.emit_from_impl([leaf, node, category], false, stream);
+                        // From<KeywordTime> for NodeEnum
+                        self.emit_from_impl([leaf, category, root], false, stream);
+                    }
+                } else {
+                    // From<Argument> for NodeEnum>
+                    self.emit_from_impl([node, category, root], false, stream);
+                }
+            }
+        }
+    }
+
+    fn emit_from_impl(
+        &self,
+        [inner, middle, outer]: [&Ident; 3],
+        inner_is_already_enum: bool,
+        stream: &mut TokenStream,
+    ) {
+        for flavor in FLAVORS {
+            let inner_ty = if inner_is_already_enum {
+                flavor.suffix_and_lifetime(inner)
+            } else {
+                flavor.ampersand(inner)
+            };
+            let middle_ty = flavor.suffix(middle);
+            let outer_ty = flavor.suffix_and_lifetime(outer);
+
+            let lifetime = (flavor != Flavor::Val).then_some(quote!(<'n>));
+
+            quote! {
+                impl #lifetime From<#inner_ty> for #outer_ty {
+                    fn from(value: #inner_ty) -> Self {
+                        Self::from(#middle_ty::from(value))
```

Comment:
is the transitive thing something that `enum_dispatch` doesn't support?

Reply:
Correct. It doesn't maintain awareness of the other enums.